### PR TITLE
Fixed #29721 -- Ensured migrations are applied and recorded atomically.

### PR DESCRIPTION
**Pull request by CuriousLearner**
cc @carltongibson 

Thanks @ubernostrum for helping during DjangoCon US sprints :)

### CODE REVIEW

**Comment by charettes** [comment]
Hey @CuriousLearner, thank you for your contribution! Do you have an idea how to create a regression test that fails without this change and passes with it? AFAIK no one managed to reproduce the issue so far.

[I asked the reporter for some form of traceback](https://code.djangoproject.com/ticket/29721#comment:1) as I couldn't see what was causing the migration to succeed without being recorded but they couldn't. While moving the record to the atomic block is technically more correct it's hard to tell if the bug doesn't lie somewhere else.

**Comment by charettes** [comment]
@gavinwahl did you happen to encounter this issue again since your initial report? Can you reproduce it systematically somehow?

**Comment by gavinwahl** [comment]
I've only encountered it naturally once. The only way I can think to reproduce it would be to put a time.sleep(5) in the right spot in django's code and then kill -9 postgres during that sleep. You might be able to accomplish the same thing in a regression test by mocking some function that's called in the right place to raise an unhandled exception.

**Comment by CuriousLearner** [comment]
This is indeed an interesting problem to solve. It is certainly not something I can think from the top of my head, but having a sleep placed in a test would increase the run-time of the whole test suite, so I'm skeptical if we would like to go with that approach.

**Comment by charettes** [comment]
Just to make it clear I wouldn't be against merging this without tests as it's definitely more technically correct and I can't think of a way to test it without mocking calls bound to implementation details.

**Comment by gavinwahl** [comment]
Maybe mock record_migrations with side_effect = AssertionError?

**Comment by charettes** [comment]
@gavinwahl I guess it's better than nothing but that wouldn't fail before the patch is applied.

**Comment by charettes** [comment]
@gavinwahl that would work but is far from your reported scenario.

I guess a check that makes sure the changes are rolled back on databases supporting transaction DLL (aka `@skipUnlessDbFeature('can_rollback_ddl')`) when `record_migration` raises an exception is better than nothing.

File: django/db/migrations/executor.py [hunk 1]
```diff
@@ -242,16 +243,23 @@ def apply_migration(self, state, migration, fake=False, fake_initial=False):
...
+                    migration_recorded = True
+        if not migration_recorded:
+            self.record_migrations(migration)
+        # Report progress
+        if self.progress_callback:
+            self.progress_callback("apply_success", migration, fake)
+        return state
+
+    def record_migrations(self, migration):
+        """Utility method to record migrations"""
```
Comment (charettes) [inline django/db/migrations/executor.py]: I think the method name is self documenting and the current docstring doesn't add much value. It should probably be named `record_migration` without an s as well.

**Comment by CuriousLearner** [comment]
Hi @charettes, thanks for your review.

I've addressed those changes. I'll work on a test, but it would take time since I'll be traveling for next week and I'm jetlagged, but I will try to work on this ASAP.

**Comment by carltongibson** [comment]
Hi @CuriousLearner. All looks good. I'm sure not but, ping me if you want/need help with the test case. 👍

**Comment by CuriousLearner** [comment]
Hi, @carltongibson @charettes 

I've tried to write a migration along with a test, but I'm not sure if it is correct. (Problem being that earlier `apply_migration` was always called and now we're making sure that it is called in the atomic block) and I'm not sure how do we ensure atomicity rollback inside the test suite.)

I did as much as I could read the code/tests and related migrations. I know this is not complete, but I would appreciate some pointers here.

Also, to add, I'm not sure why a single check in the PR is failing on djangoci jenkins as it is showing a test config error but I haven't touched any test config yet. Is it flaky, or am I missing something?

File: tests/migrations/test_executor.py [hunk 2]
```diff
@@ -649,6 +653,24 @@ def test_migrate_marks_replacement_applied_even_if_it_did_nothing(self):
             recorder.applied_migrations(),
         )
 
+    @skipIfDBFeature('can_rollback_ddl')
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_recorded_atomically"})
+    def test_migrations_applied_and_recorded_atomically(self):
+        """
+        Migrations will be applied and recorded atomically.
```
Comment (charettes) [inline tests/migrations/test_executor.py]: Migrations are applied and recorded atomically.

File: tests/migrations/test_executor.py [hunk 3]
```diff
@@ -649,6 +653,24 @@ def test_migrate_marks_replacement_applied_even_if_it_did_nothing(self):
             recorder.applied_migrations(),
         )
 
+    @skipIfDBFeature('can_rollback_ddl')
```
Comment (charettes) [inline tests/migrations/test_executor.py]: You want to run this test only if the backend has `can_rollback_ddl` instead. The reason for this is that when the feature is false the operation and the record won't be performed in a transaction and your test will systematically pass.

https://github.com/django/django/blob/58d1e9aa8ab505912389e7cd019a6f21785ad4bf/django/db/backends/base/schema.py#L86-L100

That should be `@skipUnlessDbFeature('can_rollback_ddl') instead.

File: tests/migrations/test_migrations_recorded_atomically/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+                ("name", models.CharField(max_length=255)),
+                ("slug", models.SlugField(null=True)),
+                ("age", models.IntegerField(default=0)),
+                ("silly_field", models.BooleanField(default=False)),
+            ],
+        ),
+        migrations.AlterUniqueTogether(
+            name='Author2',
+            unique_together={('name', 'slug')},
+        )
```
Comment (charettes) [inline tests/migrations/test_migrations_recorded_atomically/0001_initial.py]: No need to have such complex operations for a minimal test case; it only adds noise that future reader could interpret as necessary to cover the changes. A simple `CreateModel` with an `id` field should be enough

```python
migrations.CreateModel(
    "Foo",
    [
        ("id", models.AutoField(primary_key=True)),
    ],
)

File: tests/migrations/test_executor.py [hunk 5]
```diff
@@ -649,6 +653,24 @@ def test_migrate_marks_replacement_applied_even_if_it_did_nothing(self):
...
+        with mock.patch('django.db.migrations.executor.MigrationExecutor.record_migration') as record_migration:
+            record_migration.side_effect = RuntimeError("Recording migration failed")
+
+            with self.assertRaisesMessage(RuntimeError, "Recording migration failed"):
+                # This will call `apply_migration` which
+                # in-turn calls `record_migration`
+                executor.migrate([("migrations", "0001_initial")], fake=False)
+        # Ensure that migration record is not there
+        migration_recorder = MigrationRecorder(connection)
+        self.assertFalse(migration_recorder.migration_qs.filter(app="migrations", name="0001_initial").exists())
```
Comment (charettes) [inline tests/migrations/test_executor.py]: The important assertions here is that the operation changes were rolledback as well. You want to make sure to call `self.assertTableNotExists('migrations_foo')`.

File: tests/migrations/test_executor.py [hunk 6]
```diff
@@ -658,6 +680,7 @@ def __init__(self, graph, applied):
 
 class FakeMigration:
     """Really all we need is any object with a debug-useful repr."""
+
```
Comment (charettes) [inline tests/migrations/test_executor.py]: Revert changes.

File: tests/migrations/test_executor.py [hunk 7]
```diff
@@ -667,6 +690,7 @@ def __repr__(self):
 
 class ExecutorUnitTests(TestCase):
     """(More) isolated unit tests for executor methods."""
+
```
Comment (charettes) [inline tests/migrations/test_executor.py]: Revert changes.

File: tests/migrations/test_executor.py (hunk 6 above)
Comment (CuriousLearner) [reply tests/migrations/test_executor.py]: This was failing flake8, so I added an empty line.

Comment (charettes) [reply tests/migrations/test_executor.py]: Odd. In all cases we avoid doing unrelated style changes in such feature PR so I'd still revert.

File: tests/migrations/test_executor.py (hunk 3 above)
Comment (CuriousLearner) [reply tests/migrations/test_executor.py]: I've tried to do `from django.test import skipUnlessDBFeature` and `from django.test.testcases import skipUnlessDBFeature` but both of them are failing. I'm trying to look into this one.

File: tests/migrations/test_executor.py (hunk 5 above)
Comment (CuriousLearner) [reply tests/migrations/test_executor.py]: Done with this!

File: tests/migrations/test_migrations_recorded_atomically/0001_initial.py (hunk 4 above)
Comment (CuriousLearner) [reply tests/migrations/test_migrations_recorded_atomically/0001_initial.py]: Fixed.

**Comment by CuriousLearner** [comment]
@charettes I've addressed your review. Can you please take another pass here?

**Comment by CuriousLearner** [comment]
Awesome, thanks @charettes ! 🎉

**Comment by charettes** [comment]
Thanks for the merges @timgraham!

**Comment by CuriousLearner** [comment]
Thanks for reviewing and merging

@carltongibson on @charettes @timgraham :)

2 patches merged in a day 🎉

**Comment by charettes** [comment]
For the record this caused a significant regression when the applied migration has deferred SQL.

Now that we apply the migration and record it it in the same transaction it's possible for deferred SQL application to fail (as it's done after the transaction commits) while the migration is recorded as applied.

**Pull request merged**